toc.md

```diff
@@ -1,54 +1,76 @@
+Above is a diagram of the overall architecture of our BJP, Workerholic.
+- On the left is a stack of instances of a web application, each including Workerholic. Jobs are defined and enqueued in these web application instances.
+- In order for the jobs to be enqueued, they are first serialized and then stored in Redis, by being pushed inside Redis' List data structure, serving as `Job Queues`.
+- On the right, Workerholic workers poll from the job queues and, if there are any jobs to be done, process the jobs using the `Job Processor` component of Workerholic.
+- Regardless of whether the job is completed successfully or not, we store the job back in Redis inside a data structure serving as `Stats` storage. We will then use to this statistical information to display background jobs related metrics on our web UI.
+- If a job failed, we use the `Job Retry` component along with the `Job Scheduler` component in order to attempt to retry a job sometime in the future. To do so, a future timestamp is placed on the job, it is then stored into Redis inside a `Scheduled Jobs` sorted set (a Redis data structure that we will expand on later in this post).
+- The `Job Scheduler` will peek the sorted set and compare timestamps to see if there is a job due. If that's the case, the job will be enqueued into a Job Queue and the cycle continues.
 
 ## Building Workerholic
-Let's start diving into the numerous features we wanted in Workerholic!
+
+The idea behind this project was to build a BJP from scratch and sharing our findings, what we learned and the challenges we faced. Next, we will dive into each feature that we deemed belonged into a BJP and how they are implemented in Workerholic.
 
 ### Reliability
-A common feature developers want out of background job processors is to be reliable through a number of different reasons. It could be a network issue that prevents email sending, or maybe the job wasn't configured properly, or maybe the background job processor itself crashes. Regardless of the reason, we want to make sure that jobs are aren't just getting lost. Our challenge here is how do we make Workerholic reliable?
+
+A common feature developers want out of a BJP is to be reliable. When performing a job, a network issue that prevents email sending could occur, or the job could be misconfigured. The main application and/or the BJP could crash. Regardless of the reason, we want to make sure that jobs are not lost.
```

Comment:
I wouldn't say it's actually a feature - it's a necessity. Otherwise why would you even use it, right?

Maybe something like "One of the most important features of any background job processor is reliability."

a network issue that prevents email from being sent

Also, the main application and/or the BJP could crash.